device: binding lookup should return null if the name is null

pabigot opened the pull request:
A null device name should map to a null device.

NB: #30951 was merged to test this, but it could never have passed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

ioannisg reviewed (approved):
It looks fine to me, now, thanks @pabigot.
@nashif could we get this fix in as soon as possible? The bug was found via #31481 and currently blocks me proceeding with it :)

ioannisg commented:
> NB: #30951 was merged to test this, but it could never have passed.

#31481 should be able to catch these bugs, at least in Cortex-M. I am wondering though @andyross, I thought we have null-pointer dereferencing detection in x86, right? why this was not caught in QEMU?

nashif commented:
how is this a hotfix? is this breaking the tree (master) right now? Maybe this should be high priority bug instead?

pabigot committed:
tests: kernel: device: update to test API unconditionally

This happened to work for userspace because the syscall verifier
detected the null pointer.  Better to make sure it works in all
configurations.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
As best I can tell the test was only verifying the API when invoked from userspace.  Correct it to also check calls not from userspace.

This may be wrong, but since the test only worked because the call was from userspace something needs to be done to check this.  Perhaps the test should be built in two different ways.

nashif commented on tests/kernel/device/src/main.c:
```diff
@@ -474,8 +470,11 @@ void test_main(void)
 			 ztest_unit_test(test_dummy_device),
 			 ztest_unit_test(test_enable_and_disable_automatic_idle_pm),
 			 ztest_unit_test(test_pre_kernel_detection),
+			 ztest_unit_test(test_bogus_dynamic_name),
```
this should not needed, if userspace is not enabled, `ztest_user_unit..` behaves like  `ztest_unit...`

pabigot replied:
Yes, but the test is never run without userspace, is it?

nashif replied:
No it is not, but the fix should be in the testcase itself (which you did), not here.

ioannisg replied:
agreed

pabigot replied:
The test uses `CONFIG_TEST_USERSPACE=y` unconditionally.  AFAICT that implies `CONFIG_USERSPACE`.  The only reason the test passed at all was because it was conditional on `CONFIG_USERSPACE`, and the validation that made it work was only performed with `CONFIG_USERSPACE=y`.

So to catch the original error, it would have been necessary to both eliminate the source-code filtering of the test with `CONFIG_USERSPACE=n`, *and* to build the code with `CONFIG_USERSPACE=n`.

So from what I can tell we would need to modify this test (and potentially others) to build and run without userspace in order to catch future instances of this sort of problem.

nashif replied:
> The test uses `CONFIG_TEST_USERSPACE=y` unconditionally. AFAICT that implies `CONFIG_USERSPACE`.

all of this depends on what the platform can do, the same test can run without userspace on any platform.

ioannisg commented:
> how is this a hotfix? is this breaking the tree (master) right now? Maybe this should be high priority bug instead?

No i don't think it does, that's why I asked for a quick resolution :)

pabigot commented:
I marked it `hotfix` because at the time it wasn't clear how that test could have passed.  Now I know.

ceolin commented on kernel/device.c:
```diff
@@ -79,6 +79,10 @@ const struct device *z_impl_device_get_binding(const char *name)
 {
 	const struct device *dev;
 
+	if (name == NULL) {
```
why not use `CHECKIF` ?

ceolin reviewed (changes requested):
I'm not getting it, almost all Zephyr API's don't check for parameters because it's *assumed* that the application is trusted, we usually add an assert to the error be captured during the development. My question is why this API is different from, for instance, `k_thread_create` that does not check the pointers parameters as well ? Is this parameter coming from directly from an  untrusted source, e.g network. It does not seem to be case ...

Also,  the check in the verification function is wrong as well, AFAIK `z_user_string_copy` already handle NULL pointer.

Nevertheless, an assert check in the implementation function seems ok for me, have suggested using `CHECKIF` there.

ioannisg commented:
> Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.

@ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO

ceolin commented:
> > Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.
> 
> @ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO

The exception is gracefully handled. In this specific api when called from user thread it returns null, as expected. I didn't check other  arch but x86 though. But if I remember correctly, there are other syscalls relying on this behavior. If this is not happening we should fix there and not in device_get_binding IMHO

ioannisg commented:
> > > Also, the check in the verification function is wrong as well, AFAIK z_user_string_copy already handle NULL pointer.
> > 
> > 
> > @ceolin I am not sure I agree with this. z_user_string_copy seems to de-reference pointers, which may be NULL. I don't consider this as "handing". We should probably not de-reference pointers that are NULL, IMHO
> 
> The exception is gracefully handled. In this specific api when called from user thread it returns null, as expected. I didn't check other arch but x86 though. But if I remember correctly, there are other syscalls relying on this behavior. If this is not happening we should fix there and not in device_get_binding IMHO

Oh, I see, what you mean, let me check what is the case for Cortex-M. Thanks :)

ioannisg commented:
@ceolin I had a look and it seems Cortex-M also gracefully handles exceptions on z_user_string_copy, meaning that we do not seem to need the fix in the verification function. Won't harm, though, so I leave it up to the kernel maintainers.

pabigot commented:
Switched to use `CHECKIF()`, and removed the check in the verify function since `z_user_string_copy()` is documented to return `-EFAULT` on memory access errors.

pabigot commented:
With the merge of #32266 the problem is fixed, so now this only updates the test to be correct in a configuration that's not part of normal testing.

ceolin commented:
I'm just seeing one commit now that is removing the guards, this is wrong, passing a null pointer to `device_get_binding` without memory protection will crash.

pabigot commented:
> I'm just seeing one commit now that is removing the guards, this is wrong, passing a null pointer to `device_get_binding` without memory protection will crash.

No, it won't; see #32266.

ceolin commented:
Ok, so that that NULL check got into the tree. What is your plan now, starting to add NULL checks in all APIs ?

pabigot commented:
@ceolin is your changes-requested still relevant since the test does work correctly without usespace being enabled?

pabigot commented:
Rebased to fix the merge conflict.  (The blocking review's concerns were addressed over a month ago.)

There's been so much churn around this with the null pointer checks and trust zone, and the desire to not explicitly check for failures but instead rely on exceptions to detect problems, I honestly don't know if this is a correct fix anymore.  I note that `z_user_string_nlen` still documents its behavior as undefined in the case of error, which seems like a bad idea.

LeiW000 commented:
@nashif , it looks that the PR can be merged now.
